### PR TITLE
[Capture] Qfunc jaxpr is cached when executing `QNode`s

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
This PR adds support for caching the jaxpr of a qnode's qfunc. Additionally, this PR adds support in `QNode` to be able to specify `static_argnums`.

**Description of the Change:**
* Add `capture_cache` attribute to `QNode`. This is an `LRUCache` that is supposed to hold the jaxpr for the qfunc for various argument/keyword argument types.
* Add `static_argnums` argument to `QNode` for tracking `static_argnums`
* Add `_split_static_args` helper function to `pennylane/workflow/capture_qnode.py` to separate the dynamic and static arguments of a QNode.
* Add `_get_jaxpr_cache_key` helper function to `pennylane/workflow/capture_qnode.py` to compute the hash used as a key for the jaxpr cache. This key depends on the abstract eval (shape + dtype) of the positional arguments, and the concrete values of the keyword arguments.
  * If `static_argnums` are provided, we use the concrete eval of the positional arguments to compute the hash.
  * If any of the arguments have dynamic shapes, the letter(s) corresponding to the abstract axis is used in the shape tuple of the serialized argument.

**Benefits:**
* We can cache jaxpr, eliminating the need to re-compute jaxpr every time a qnode is executed.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

[sc-83774] [sc-84375]

### CODE REVIEW

File: tests/capture/workflow/test_capture_qnode.py [hunk 1]
```diff
@@ -1218,3 +1217,437 @@ def circuit():
             return qml.state()
 
         assert qml.math.allclose(circuit(), [0, 0, 0, 1])
+
+
+def test_qnode_static_argnums():
```
Comment (lillian542) [inline tests/capture/workflow/test_capture_qnode.py:1222]: Can we add a test that would have failed before you added the `sort` to `qnode.static_argnums`?

**Review by lillian542 — APPROVED** [review]
One non-blocking suggestion, looks good to me! 🚀

**Review by andrijapau — COMMENTED** [review]
Nice! 🚀 I just have a few comments that probably come from a lack of experience. 😆

File: pennylane/workflow/qnode.py [hunk 2]
```diff
@@ -604,6 +607,11 @@ def __init__(
         _validate_diff_method(self.device, self.diff_method)
         cache = (max_diff > 1) if cache == "auto" else cache
 
+        self.capture_cache = LRUCache(maxsize=1000)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/qnode.py:610]: I'll let @isaacdevlugt confirm, but for now I think the goal was to just have this feature "work". Users don't really need to have a way to interface with it right now. Though I see no reason why we can't/shouldn't use `cache_size`.

File: tests/capture/workflow/test_capture_qnode.py [hunk 3]
```diff
@@ -1218,3 +1217,437 @@ def circuit():
...
+        # Cache hit because same arg shape/type and same static arg
+        spy.reset_mock()
+        self.check_execution_results(circuit, 1.1, 2.0)
+        spy.assert_not_called()
+
+        # Cache miss because same arg shape/type but different static arg
+        spy.reset_mock()
+        self.check_execution_results(circuit, 1.1, 2.1)
+        spy.assert_called()
+
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/capture/workflow/test_capture_qnode.py:1432]: Nope, won't be any different. For dynamic args, we don't need to worry about the value, just shape/dtype. I'll add this to the test regardless though. Doesn't hurt :)

File: tests/capture/workflow/test_capture_qnode.py [hunk 4]
```diff
@@ -1218,3 +1217,437 @@ def circuit():
...
+            (jnp.array(2.5), 2.5),
+            (jnp.array(1.5), 3.5),
+            (jnp.array(2.5), 3.5),
+        ]
+
+        for a in args:
+            res = jitted_circuit(*a)
+
+            with qml.capture.pause():
+                assert qml.math.allclose(res, circuit(*a))
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/capture/workflow/test_capture_qnode.py:1653]: The only real difference (other than the intention of the respective tests) is that we're verifying the correctness of the results *when we know that the cached jaxpr should be getting used*. I guess that we don't really need `test_qnode_static_argnums_jit`. I'll remove it and leave a comment.

File: tests/capture/workflow/test_capture_qnode.py [hunk 5]
```diff
@@ -1218,3 +1217,437 @@ def circuit():
...
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RX(jnp.sum(x), 0)
+            return qml.expval(qml.Z(0))
+
+        spy = mocker.spy(jax, "make_jaxpr")
+        _ = jax.make_jaxpr(circuit, abstracted_axes=("a",))(jnp.arange(10))
+        assert spy.call_count > 1
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/capture/workflow/test_capture_qnode.py:1475]: We don't know how many times it might get called. Not specifically for this test, but in general, it will likely get called multiple times through different interpreters, when creating jaxpr for higher order primitives, etc. However, we know for certain how many times it should _not_ get called when caching is active. The intention of this test is not to count how many times we call `jax.make_jaxpr`, but rather to verify that we're _not_ calling it when we expect the cache to get used.

[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/capture/workflow/test_capture_qnode.py:1475]: Ah I see, okay just wanted to make sure.

File: pennylane/workflow/_capture_qnode.py [hunk 6]
```diff
@@ -428,18 +524,21 @@ def f(x):
...
         device=qnode.device,
         qnode_kwargs=qnode_kwargs,
         qfunc_jaxpr=qfunc_jaxpr.jaxpr,
         n_consts=len(qfunc_jaxpr.consts),
     )
-    assert flat_fn.out_tree is not None, "out_tree should be set by call to flat_fn"
-    return jax.tree_util.tree_unflatten(flat_fn.out_tree, res)
+
+    if not using_cached_plxpr:
+        assert flat_fn.out_tree is not None, "out_tree should be set by call to flat_fn"
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/_capture_qnode.py:540]: When we're not using the cache, `out_tree` should _never_ be `None`, so, this is more of a dev-facing assertion to immediately check if something has gone wrong. It should never be hit in a normal workflow. I can change it if you'd like, but it shouldn't impact user experience whatsoever.

File: tests/capture/workflow/test_capture_qnode.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/capture/workflow/test_capture_qnode.py:1653]: Got it!

File: pennylane/workflow/_capture_qnode.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/workflow/_capture_qnode.py:540]: Hmm, I see. If there is absolutely no situation where `flat_fn.out_tree = None` then I guess we can leave it (but then I wonder why have the assertion in the first place). 

Wouldn't raising a `CaptureError` be a happy middle ground? Up to you ... non-blocking for me. 👍🏼

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/_capture_qnode.py:540]: I'll update it. Agree that it doesn't hurt.

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/_capture_qnode.py:540]: I'd prefer to leave it as is. There are 7 places in the source code with the same assertion. I don't think it's a particularly problematic thing if we raise an `AssertionError` instead of a `CaptureError`.

**Review by andrijapau — APPROVED** [review]
Nice job! 🚀 I guess all we are waiting for is @isaacdevlugt's reply.

**Review by lillian542 — APPROVED** [review]
🚀

File: pennylane/workflow/qnode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaacdevlugt) [reply pennylane/workflow/qnode.py:610]: Hey! I would say that users probably never need to interact with a captured object's cache size. Developers, sure. So, whatever works from your guys' point of view as developers is fine by me :)

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/workflow/qnode.py:610]: I'll leave it as is for now. It won't be difficult to update it if we decide to do so. Only 2-3 lines to change.
